kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala

```diff
@@ -3,6 +3,59 @@ package me.rexim.morganey.ast
 import me.rexim.morganey.church.ChurchNumberConverter._
 import me.rexim.morganey.church.ChurchPairConverter._
 
+import scala.annotation.tailrec
+
+object LambdaTerm {
+
+  private def nestedApplications(app: LambdaApp): (LambdaTerm, List[LambdaTerm]) = {
+    @tailrec
+    def go(app: LambdaApp, acc: List[LambdaTerm]): (LambdaTerm, List[LambdaTerm]) =
+      app match {
+        case LambdaApp(l: LambdaApp, r) => go(l, r :: acc)
+        case LambdaApp(l, r)            => (l, r :: acc)
+      }
+    go(app, Nil)
+  }
+
+  private def nestedFunctions(func: LambdaFunc): (List[LambdaVar], LambdaTerm) = {
+    @tailrec
+    def go(func: LambdaFunc, acc: List[LambdaVar]): (List[LambdaVar], LambdaTerm) =
+      func match {
+        case LambdaFunc(v, f: LambdaFunc) => go(f, v :: acc)
+        case LambdaFunc(v, b)             => ((v :: acc).reverse, b)
+      }
+    go(func, Nil)
+  }
+
+  private def hasAppAsBody(func: LambdaFunc): Boolean =
+    nestedFunctions(func) match {
+      case (_, _: LambdaApp) => true
+      case _                 => false
+    }
+
+  def prettyPrint(term: LambdaTerm): String = term match {
+    case LambdaVar(name)  => name
+    case _: LambdaInput   => "<input>"
+    case app: LambdaApp   =>
+      val (deep, nest) = nestedApplications(app)
+      val nested = nest.map {
+        case a: LambdaApp                     => s"(${prettyPrint(a)})"
+        case f: LambdaFunc if hasAppAsBody(f) => s"(${prettyPrint(f)})"
+        case t                                => prettyPrint(t)
+      }.mkString(" ")
+      val rest = deep match {
+        case f: LambdaFunc => s"(${prettyPrint(f)})"
+        case t             => prettyPrint(t)
+      }
+      s"$rest $nested"
+    case func: LambdaFunc =>
+      val (vars, body) = nestedFunctions(func)
+      val variables = vars.map(t => s"${prettyPrint(t)}.").mkString("")
+      s"λ$variables${prettyPrint(body)}"
+  }
```

Comment:
> I mean you could take each branch of this match-case expression and put them into toString-s of corresponding classes that you are trying to match.

You are absolutely right with that. First I thought the implementations of each branch had to be more intermingled, so I started implementing it as a static function. I just kept it that way.
This is how I prefer it personally.

> I have to also not forget to add another branch to this expression which is in a different place for some reason. 

The Scala compiler will give you a warning for the not exhaustive pattern match.

--
If you prefer the polymorphic function I'll gladly split each cases and move them into the classes.